src/qibo/core/terms.py

```diff
@@ -0,0 +1,273 @@
+import sympy
+from qibo import gates, K
+from qibo.config import raise_error, EINSUM_CHARS
+
+
+class HamiltonianTerm:
+    """Single term in a :class:`qibo.core.hamiltonians.SymbolicHamiltonian`.
+
+    This Hamiltonian is represented by a sum of such terms which are held in
+    its ``.terms`` attribute which is a list.
+
+    Args:
+        matrix (np.ndarray): Full matrix corresponding to the term representation
+            in the computational basis. Has size (2^n, 2^n) where n is the
+            number of target qubits of this term.
+        q (int): Target qubit ids.
+    """
+
+    def __init__(self, matrix, *q):
+        self.target_qubits = tuple(q)
+        self._gate = None
+        self.hamiltonian = None
+        if not (matrix is None or isinstance(matrix, K.qnp.numeric_types) or
+                isinstance(matrix, K.qnp.tensor_types)):
+            raise_error(TypeError, "Invalid type {} of symbol matrix."
+                                   "".format(type(matrix)))
+        self._matrix = matrix
+
+    @property
+    def matrix(self):
+        return self._matrix
+
+    @property
+    def gate(self):
+        """Qibo gate that implements the action of the term on states."""
+        if self._gate is None:
+            self._gate = gates.Unitary(self.matrix, *self.target_qubits)
+        return self._gate
+
+    def exp(self, dt):
+        """Matrix exponentiation of the term."""
+        return K.qnp.expm(-1j * dt * self.matrix)
```

Comment:
Is this correct? Or should we drop the `qnp`?

Reply:
Check my comment below regarding the K vs K.qnp distinction. I think leaving K.qnp here would generally be faster because we are exponentiating small matrices (2x2 or 4x4).

Also this is an example case where K.qnp has different behavior from K.np. K.np would not work here because numpy does not provide an `expm` method and we have to use the one from scipy (which is available via `K.qnp`).